pallets/dmp-queue/src/lib.rs

```diff
@@ -0,0 +1,782 @@
+
+/// Index used to identify normal pages.
+pub type PageCounter = u32;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// The module configuration trait.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type XcmExecutor: ExecuteXcm<Self::Call>;
+
+		/// Origin which is allowed to execute overweight messages.
+		type ExecuteOverweightOrigin: EnsureOrigin<Self::Origin>;
+	}
+
+	/// The configuration.
+	#[pallet::storage]
+	pub(super) type Configuration<T> = StorageValue<_, ConfigData, ValueQuery>;
+
+	/// The page index.
+	#[pallet::storage]
+	pub(super) type PageIndex<T> = StorageValue<_, PageIndexData, ValueQuery>;
+
+	/// The queue pages.
+	#[pallet::storage]
+	pub(super) type Pages<T> = StorageMap<
+		_,
+		Blake2_128Concat,
+		PageCounter,
+		Vec<(RelayBlockNumber, Vec<u8>)>,
+		ValueQuery,
+	>;
+
+	/// The overweight messages.
+	#[pallet::storage]
+	pub(super) type Overweight<T> = StorageMap<
+		_,
+		Blake2_128Concat,
+		OverweightIndex,
+		(RelayBlockNumber, Vec<u8>),
+		OptionQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The message index given is unknown.
+		Unknown,
+		/// The amount of weight given is possibly not enough for executing the message.
+		OverLimit,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_now: T::BlockNumber, max_weight: Weight) -> Weight {
+			// on_idle processes additional messages with any remaining block weight.
+			Self::service_queue(max_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Service a single overweight message.
+		///
+		/// - `origin`: Must pass `ExecuteOverweightOrigin`.
+		/// - `index`: The index of the overweight message to service.
+		/// - `weight_limit`: The amount of weight that message execution may take.
+		///
+		/// Errors:
+		/// - `Unknown`: Message of `index` is unknown.
+		/// - `OverLimit`: Message execution may used greater than `weight_limit`.
+		///
+		/// Events:
+		/// - `OverweightServiced`: On success.
+		#[pallet::weight(1_000_000 + weight_limit)]
+		pub fn service_overweight(
+			origin: OriginFor<T>,
+			index: OverweightIndex,
+			weight_limit: Weight,
+		) -> DispatchResultWithPostInfo {
+			T::ExecuteOverweightOrigin::ensure_origin(origin)?;
+
+			let (sent_at, data) = Overweight::<T>::get(index).ok_or(Error::<T>::Unknown)?;
+			let used = Self::try_service_message(weight_limit, sent_at, &data[..])
+				.map_err(|_| Error::<T>::OverLimit)?;
+			Overweight::<T>::remove(index);
+			Self::deposit_event(Event::OverweightServiced(index, used));
+			Ok(PostDispatchInfo { actual_weight: Some(1_000_000 + used), pays_fee: Pays::Yes })
```

Comment:
```suggestion
			Ok(Some(1_000_000 + used).into())
```
should work I think?